No experimental check for platform setting [MYC-117]

tommyknows opened the pull request:
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?

Mainly removes the check if Docker experimental is enabled if the user specifies a platform. It's not needed anymore as the `--platform` flag for `docker pull` has been graduated to stable. Also some clean-up around that feature.

#### Any background context you want to provide?

#### What are the relevant tickets?

https://snyksec.atlassian.net/browse/MYC-117

yaronschwimmer commented:
[MYC-131](https://snyksec.atlassian.net/browse/MYC-131) is definitely not the relevant ticket. Is there a relevant ticket?

tommyknows commented:
> [MYC-131](https://snyksec.atlassian.net/browse/MYC-131) is definitely not the relevant ticket. Is there a relevant ticket?

Sorry, that was a copy-paste issue 😅 Yes, [MYC-117](https://snyksec.atlassian.net/browse/MYC-117). PR description updated as well. 

I'm not sure about the second commit, I marked it as a `BREAKING CHANGE: ...`, but it does not appear in the expected release notes? 🤔

yaronschwimmer commented:
> > [MYC-131](https://snyksec.atlassian.net/browse/MYC-131) is definitely not the relevant ticket. Is there a relevant ticket?
> 
> Sorry, that was a copy-paste issue 😅 Yes, [MYC-117](https://snyksec.atlassian.net/browse/MYC-117). PR description updated as well.
> 
> I'm not sure about the second commit, I marked it as a `BREAKING CHANGE: ...`, but it does not appear in the expected release notes? 🤔

I don't think it's a breaking change, as it doesn't change anything in the API of the library. The expected release notes is not great with identifying this I think. Why did you consider this a breaking change? 
Also, I wonder why there are 3 commits in this rather small PR. Does each of these commits stand on its own?

yaronschwimmer commented on lib/docker.ts:
```diff
@@ -51,24 +51,13 @@ class Docker {
-
   public async pullCli(targetImage: string, options?: DockerOptions) {
-    let platformOption: string = "";
+    const opts: string[] = ["pull", targetImage];
```
I'm not sure how smart Docker is with adding the option _after_ the image name (it might, I just don't know and want to make sure it works)

yaronschwimmer commented on lib/docker.ts:
```diff
@@ -51,24 +51,13 @@ class Docker {
-          '"--platform" is only supported on a Docker daemon with experimental features enabled',
-        );
-      }
+      opts.push(`--platform=${options.platform}`);
```
I'm not sure how smart Docker is with adding the option after the image name (it might be ok, I just don't know and want to make sure it works)

tommyknows replied:
It does work, tested that locally (honestly, this should work for every sensible CLI imo, but that's another topic 😄). 

Not sure if it makes sense to test that though? And if yes, how?

yaronschwimmer commented:
> > Why did you consider this a breaking change?
> 
> It removes a function that had been marked as `public`. Even though I can't find any other usages, this is a breaking change, right? 😅
> 
> > Also, I wonder why there are 3 commits in this rather small PR. Does each of these commits stand on its own?
> 
> They certainly could. Whether it would make sense or not... 🤷‍♂️
> 
> I can also squash them into a single commit. Is there some guidance on when to split up / squash specific commits / changes? I put them in three separate commits to more easily describe them, but obviously I can also do the same in a single, larger commit message :)

Well if nothing uses it internally, and it's not exported as part of the library's API (essentially everything that's exported in `lib/index.ts`), and it doesn't change the output of any exported function, then it's not a breaking change IMO.
As for the commits, there are no strict guidelines. If they make sense as separate commits (in the sense that you can theoretically revert one without having to revert the others), then leave it as it is.

tommyknows commented:
> Well if nothing uses it internally, and it's not exported as part of the library's API (essentially everything that's exported in lib/index.ts), and it doesn't change the output of any exported function, then it's not a breaking change IMO.

Ah I did not know that, I thought a `public ...` was...public :-) Will need to read up on that and change the commit.

> As for the commits, there are no strict guidelines. If they make sense as separate commits (in the sense that you can theoretically revert one without having to revert the others), then leave it as it is.

Theoretically with some merge conflicts you probably could, but it would probably not make too much sense :smile:

tommyknows commented:
I dropped / squashed the breaking-change commit into the `remove docker experimental check` commit, but decided to keep the refactor-commit separate.

yaronschwimmer reviewed (approved):
Congratulations on your first PR at Snyk 🎉

tommyknows commented on test/lib/analyzer/image-inspector.spec.ts:
```diff
@@ -182,6 +182,82 @@ describe("getImageArchive", () => {
+    });
+  });
+
+  describe("from remote registry with experimental platform feature binary", () => {
```
This and the above new tests are pretty much the exact same apart from the returned errors.
Should I use a loop instead for this? 
Or only test one of these?

tommyknows commented:
Updated the PR with the things we discussed in Slack.

Note that this doesn't really change behaviour in cases of errors: we still *do not* fall back to pulling with the pull library. 
In the two cases that I've added it would not make sense anyway as the pull library does not support a platform setting anyways.

tommyknows commented:
@yaronschwimmer may I ask for another review please? Thanks!

yaronschwimmer reviewed (approved):
🎉

snyksec commented:
:tada: This PR is included in version 5.1.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/snyk-docker-plugin/v/5.1.0)
- [GitHub release](https://github.com/snyk/snyk-docker-plugin/releases/tag/v5.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
